examples/reference/custom_components/ReactiveESM.md

```diff
@@ -0,0 +1,283 @@
+            border-radius:4px;
+        }
+        button:hover {
+            background: #4099da;
+        }
+        """
+    ]
+
+    _esm = """
+    export function render({ data }) {
+        let btn = document.createElement("button");
+        btn.innerHTML = `count is ${data.value}`;
+        btn.addEventListener("click", () => {
+            data.value+=1
+        });
+        data.watch(() => {
+            btn.innerHTML = `count is ${data.value}`;
+          }, 'value')
+        return btn
+    }
+    """
+
+StyledCounterButton().servable()
+```
+
+## Send Events from Javascript to Python
+
+You can send events from JavaScript to Python using the `data.send_event` method. You can handle the event in Python by defining an `on_<event_name>` method.
+
+```python
+import panel as pn
+import param
+
+pn.extension()
+
+class ButtonEventExample(pn.ReactiveESM):
+
+    value = param.Parameter()
+
+    _esm = """
+    export function render({ data }) {
+        const btn = document.createElement('button')
+        btn.innerHTML = `Click Me`;
+        btn.onclick = (event) => data.send_event('click', event)
+        return btn
+    }
+    """
+
+    def on_click(self, event):
+        self.value=str(event.__dict__)
+
+button = ButtonEventExample()
+pn.Column(
+    button, pn.widgets.TextAreaInput(value=button.param.value, height=200),
+).servable()
+```
+
+## Dependency Imports
+
+JavaScript dependencies can be imported directly via a fully qualified URL for example from [`esm.sh`](https://esm.sh/).
+
+```python
+import panel as pn
+
+pn.extension()
+
+class ConfettiButton(pn.ReactiveESM):
+
+    _esm = """
+    import confetti from "https://esm.sh/canvas-confetti@1.6.0";
+
+    export function render() {
+        let btn = document.createElement("button");
+        btn.innerHTML = `Click Me`;
+        btn.addEventListener("click", () => {
+            confetti()
+        });
+        return btn
+    }
+    """
+
+ConfettiButton()
+```
+
+We can use the `_import_map` attribute to user shorter module names when importing
+
+```python
+import panel as pn
+
+pn.extension()
+
+class ConfettiButton(pn.ReactiveESM):
+    _import_map = {
+        "canvas-confetti": "https://esm.sh/canvas-confetti@1.6.0",
+    }
+
+    _esm = """
+    import confetti from "canvas-confetti";
+
+    export function render() {
+        let btn = document.createElement("button");
+        btn.innerHTML = `Click Me`;
+        btn.addEventListener("click", () => {
+            confetti()
+        });
+        return btn
+    }
+    """
+
+ConfettiButton()
+```
+
+## External Files
+
+You can also load the JavaScript from an external file by providing a path to the `_esm` attribute.
+
+Create the file **counter_button.py**.
+
+```python
+from pathlib import Path
+
+import param
+
+import panel as pn
+
+pn.extension()
+
+
+class CounterButton(pn.ReactiveESM):
+
+    value = param.Integer()
+
+    _esm = Path("counter_button.js")
```

Comment:
We should consider supporting strings pointing to files just as AnyWidget does `_esm = "counter_button.js"`.

Reply:
No objection, it'll just assume it's relative to the component itself.

Reply:
Done